File: algorithms/add_two_numbers.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
  l = ListNode.new(0)
  head = l
  r = 0
  while l1 or l2 or r != 0
    a = b = 0
    a = l1.val if l1
    b = l2.val if l2
    s = a + b + r
    l.val = s % 10
    r = s / 10
    pre = l
    l = ListNode.new(0)
    pre.next = l
    l1 = l1.next if l1
    l2 = l2.next if l2
  end
  pre.next = nil
  head
end

def create_test_list(*digits)
  l = ListNode.new(0)
  head = pre = l
  digits.each do |digit|
    l.val = digit
    pre = l
    pre.next = l = ListNode.new(0)
  end
  pre.next = nil
  head
end

l1 = create_test_list(2, 4, 3)
l2 = create_test_list(5, 6, 4)

p add_two_numbers(l1, l2)
p add_two_numbers(ListNode.new(4), ListNode.new(6))

File: algorithms/add_two_numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        dummy = p = ListNode(None)
        r = 0
        while l1 or l2 or r:
            p.next = ListNode(0)
            p = p.next

            a = b = 0

            if l1:
                a = l1.val
                l1 = l1.next

            if l2:
                b = l2.val
                l2 = l2.next

            s = a + b + r
            p.val = s % 10
            r = s // 10

        return dummy.next
